### PR TITLE
feat: Polars generic dataset (take 3)

### PR DESCRIPTION
**Pull request by astrojuanlu**
## Description
Close gh-110.

## Development notes
See past discussion in gh-153 and gh-116.

Not 100 % sure I did the rebase right, will look into it later.

## Checklist

- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the relevant `RELEASE.md` file
- [ ] Added tests to cover my changes

### CODE REVIEW

File: kedro-datasets/RELEASE.md
Comment (jmholzer) [inline kedro-datasets/RELEASE.md:6]:
```suggestion
* Added `polars.GenericDataSet`, a `GenericDataSet` backed by [polars](https://www.pola.rs/), a lightning fast dataframe package built entirely using Rust.
```

**Comment by astrojuanlu** [comment]
Lots of test failures that are completely unrelated to this PR. Will have a look later.

**Comment by sbrugman** [comment]
This dataset would be of great use! Happy to help where possible.

The error that is mostly present is:
```
TypeError: expected exception must be a BaseException type, not DataSetError
```

The kedro version compared to main shows changes in `kedro.io.core`:

https://github.com/kedro-org/kedro/compare/08d3a020081602208180897ec38e1f400dd24f0b...main#diff-236fdc9ebc4cf9a942f1044bec507416ca1ab60918d41177e04b18591fecb40d

This probably has to do with the renaming of `DataSetError` to `DatasetError`:

kedro/io/core.py

_kedro/io/core.py_

I'm not well-known with the transition plans for the `DataSet` names and the parity of `kedro` and `kedro-datasets`. Do you think it's possible to move forward?

**Comment by astrojuanlu** [comment]
Hi @sbrugman, that help would be really appreciated! This has been on my to-do list for too long.

I just rebased and force pushed on top of current `main`, let's see what the CI says.

**Comment by astrojuanlu** [comment]
About the move to `*Dataset` from `*DataSet`, https://github.com/kedro-org/kedro-plugins/pull/255 should take care of that I believe. But it's a bit stuck.

**Comment by sbrugman** [comment]
LGTM!

**Comment by astrojuanlu** [comment]
Looks like everything is green now 😄 

Have you used it already @sbrugman? To be honest I haven't looked at the code nor tried to use it yet myself. But it looks like

- `kedro_datasets/polars/generic_dataset.py` has 100 % coverage
- Tests are parametrized by file type and location

I'm... marking this as ready for review 😬

**Comment by sbrugman** [comment]
As it stands, we will use it for parquet on S3 once this dataset is incorporated. Will report issues if we run into them.

**Comment by astrojuanlu** [comment]
I tested this with https://github.com/astrojuanlu/workshop-jupyter-kedro successfully:

```diff
diff --git a/conf/base/catalog.yml b/conf/base/catalog.yml
index 854b5e8..f06dd7e 100644
--- a/conf/base/catalog.yml
+++ b/conf/base/catalog.yml
@@ -12,22 +12,14 @@ openrepair-0_3-categories:
   filepath: data/01_raw/OpenRepairData_v0.3_Product_Categories.csv
 
 openrepair-0_3-combined:
-  type: polars.CSVDataSet
-  filepath: data/02_intermediate/openrepairdata_v0.3_combined.csv
-  load_args:
-    dtypes:
-      product_age: ${pl:Float64}
-      group_identifier: ${pl:Utf8}
-    try_parse_dates: true
+  type: polars.GenericDataSet
+  file_format: parquet
+  filepath: data/02_intermediate/openrepairdata_v0.3_combined.pq
 
 openrepair-0_3:
-  type: polars.CSVDataSet
-  filepath: data/03_primary/openrepairdata_v0.3_clean.csv
-  load_args:
-    dtypes:
-      product_age: ${pl:Float64}
-      group_identifier: ${pl:Utf8}
-    try_parse_dates: true
+  type: polars.GenericDataSet
+  file_format: parquet
+  filepath: data/03_primary/openrepairdata_v0.3_clean.pq
 
 wordcloud-plot:
   type: matplotlib.MatplotlibWriter
```

and everything worked perfectly. I cannot approve my own pull request 😬 so, reviews welcome!

**Comment by astrojuanlu** [comment]
> Why do we need to keep bumping the lower bound? Is there something that prevents us from just using a less-conservative upper bound (and leaving a minimal lower bound)?

_Originally posted by @deepyaman in https://github.com/kedro-org/kedro-plugins/pull/268#pullrequestreview-1530333418_

Reposting it here because it's relevant. Maybe I can keep the lower bound.

**Comment by astrojuanlu** [comment]
Tagging this as `Community` because actually I didn't write most of this PR, I just kept updating it 😄

**Comment by datajoely** [comment]
@astrojuanlu shall we get this in?

**Comment by astrojuanlu** [comment]
@datajoely needs approvals 🥲

**Comment by astrojuanlu** [comment]
I tested this on my Kedro + Polars workshop material and works perfectly

**Review by datajoely — APPROVED** [review]

File: kedro-datasets/tests/polars/test_generic_dataset.py
Comment (merelcht) [inline kedro-datasets/tests/polars/test_generic_dataset.py:117]: I would change this to:
```suggestion
    def test_assert_write_mode_fails(self):
```
in line with the tests below that have fail in the name for a test that should raise an exception.

File: kedro-datasets/kedro_datasets/polars/generic_dataset.py
Comment (astrojuanlu) [reply kedro-datasets/kedro_datasets/polars/generic_dataset.py:210]: Trying to understand this too. From the docstring:

> Accepted values are "overwrite" and "ignore", use "ignore" when you want to read fileformat that polars does not provide write support for.

So, you're right: this logic is weird. Should we just remove the `write_mode` parameter and replace this chunk by

```python
        if self._file_format not in ACCEPTED_WRITE_FILE_FORMATS:
            raise DataSetError(
                f"Unable to retrieve 'polars.DataFrame.write_{self._file_format}' "
                "method, please "
                "ensure that your 'file_format' parameter has been defined correctly as"
                " per the Polars API "
                "https://pola-rs.github.io/polars/py-polars/html/reference/io.html"
            )
```

?

File: kedro-datasets/kedro_datasets/polars/generic_dataset.py
Comment (astrojuanlu) [reply kedro-datasets/kedro_datasets/polars/generic_dataset.py:79]: Yeah it's weird that "write_mode" is touted as something you need to change if you want to _read_ a specific format. See my comment above.

File: kedro-datasets/kedro_datasets/polars/generic_dataset.py
Comment (merelcht) [reply kedro-datasets/kedro_datasets/polars/generic_dataset.py:210]: Yes that makes more sense. And then should the call to `self._ensure_file_system_target()` in line 220 be removed as well? That just checks `ACCEPTED_READ_FILE_FORMATS`. Perhaps `_ensure_file_system_target()` doesn't even need to be a separate method.

File: kedro-datasets/RELEASE.md
Comment (ankatiyar) [inline kedro-datasets/RELEASE.md:5]: This needs to be updated!

**Review by ankatiyar — APPROVED** [review]
I tested it too and it works fine! :)

File: kedro-datasets/kedro_datasets/polars/generic_dataset.py
Comment (SajidAlamQB) [inline kedro-datasets/kedro_datasets/polars/generic_dataset.py:162]:
```suggestion
        if not load_method:
            raise DataSetError(
                f"Unable to retrieve 'polars.read_{self._file_format}' method, please"
                " ensure that your "
                "'file_format' parameter has been defined correctly as per the Polars"
                " API"
                " https://pola-rs.github.io/polars/py-polars/html/reference/io.html"
            )
        with self._fs.open(load_path, **self._fs_open_args_load) as fs_file:
            return load_method(fs_file, **self._load_args)

```

Comment (SajidAlamQB) [reply kedro-datasets/kedro_datasets/polars/generic_dataset.py:162]: minor nitpick: I don't think the `else` is needed since if `DataSetError` is raised it will exit this flow without reaching it.

File: kedro-datasets/kedro_datasets/polars/generic_dataset.py
Comment (SajidAlamQB) [reply kedro-datasets/kedro_datasets/polars/generic_dataset.py:180]: Same as above comment.

File: kedro-datasets/RELEASE.md
Comment (astrojuanlu) [reply kedro-datasets/RELEASE.md:3]: Done!

**Comment by astrojuanlu** [comment]
Polars 0.19 is out 😄 after this PR is merged I'll verify that nothing broke here https://github.com/pola-rs/polars/releases/tag/py-0.19.0

**Review by merelcht — APPROVED** [review]
This looks great! Let's get it merged 👍

**Comment by astrojuanlu** [comment]
🔥 Thanks a lot reviewers! Super happy to see this land 👏🏽

**Comment by astrojuanlu** [comment]
And thanks a lot to @wmoreiraa for starting the original effort!

**Comment by grofte** [comment]
Oh, I thought this was already available in 0.19.1. If I add a pandas<->polars conversion at the beginning and end of every node I could have a polars pipeline, right? And then switch over once a new version is released that reads and writes Polars?

**Comment by astrojuanlu** [comment]
@grofte This is available in kedro-datasets! You have to `pip install kedro-datasets[polars]`

**Comment by grofte** [comment]
I'm running Kedro 0.19.2 and I've done `pip install kedro-dataset[polars]` (and repeated it after uninstalling kedro-dataset). The kedro-dataset version is 2.1.0.

I kept getting `kedro.io.core.DatasetError: Class 'polars.GenericDataSet' not found, is this a typo?` but then I found out that S->s so now I get `Class 'polars.GenericDataset' not found, is this a typo?`. According to this https://docs.kedro.org/projects/kedro-datasets/en/kedro-datasets-2.0.0/api/kedro_datasets.html GenericDataset is only a thing for Pandas.

**Comment by astrojuanlu** [comment]
Oh, yes it changed its name I believe 🙏🏽 Glad you got it to work! If you have any troubles with it, feel free to open an issue.

**Comment by grofte** [comment]
Thank you!

And update your blog post :wink: https://kedro.org/blog/a-polars-exploration-into-kedro

It's the premiere source for kedro+polars information :grin:

**Comment by grofte** [comment]
For any future folks that come across this: There's a bug, I think in Polars but maybe in Kedro Datasets, where Kedro's EagerDataset opens the parquet file but polars doesn't recognize it as bytes / io.BufferedIOBase / io.RawIOBase and therefore sends it to `.scan_parquet` which only takes paths as an argument. I don't know which library it is happening in, or why, so I can't make a PR to fix it but there's a work-around. In your `catalog.yml` add
```yaml
load_args:
    use_pyarrow: true
```

**Comment by astrojuanlu** [comment]
Ugh, could you open a separate issue about this @grofte ?

**Pull request merged**
